src/services/youtube/api.ts

```diff
@@ -396,20 +397,34 @@ class YoutubeClient implements IYoutubeApi {
     }
   }
 
+  getNextProxy() {
+    const proxies = this.config.proxy?.urls
+    if (!proxies) {
+      return undefined
+    }
+    ++this.proxyIdx
+    if (this.proxyIdx >= proxies.length) {
+      this.proxyIdx = 0
+    }
+    return proxies[this.proxyIdx]
+  }
+
   async downloadVideo(videoUrl: string, outPath: string): ReturnType<typeof ytdl> {
     // Those flags are not currently recognized by `youtube-dl-exec`, but they are still
     // supported by yt-dlp (see: https://github.com/yt-dlp/yt-dlp)
+    const proxy = this.getNextProxy()
+    console.error(`Using proxy ${proxy} to download ${videoUrl}...`)
     const response = await ytdl(videoUrl, {
       noWarnings: true,
       printJson: true,
       format: 'bv*[height<=1080][ext=mp4]+ba[ext=m4a]/bv*[height<=1080][ext=webm]+ba[ext=webm]/best[height<=1080]',
       output: `${outPath}/%(id)s.%(ext)s`,
       ffmpegLocation: ffmpegInstaller.path,
       // forceIpv6: true,
-      limitRate: '4M',
+      limitRate: this.config.sync.limits?.bandwidthPerDownload,
```

Comment:
In my testing I didn't find the `limitRate` effectively limited the rate to desired value. It is the main metric by which we get blocked. And it blocks entire network range, not a single IP so we should to be sure.

Reply:
Thank you for pointing that out, I changed the approach to use `proxychains` + `tc` for rate limiting instead.